versioned_hdf5/wrappers.py

```diff
@@ -492,6 +493,141 @@ def _make_new_dset(
     return data
 
 
+def as_subchunk_map(chunk_size: ChunkSize, idx, shape: tuple):
```

Comment:
I'm still trying to wrap my head around what this does and how it differs from `as_subchunks`, but this looks like pure index manipulation so it's probably in scope for adding directly to ndindex if you want.

Reply:
Most of the indexing operations in `versioned-hdf5` are some kind of data access where we either need to read or write to a chunk in some larger dataset. Practically speaking that means that basically every time we have a read/write to some `index`, we

1. loop over each `chunk` that get hit by `index` using `ChunkSize(chunk_size).as_subchunks(index, shape)`,
2. get the indices of each subchunk that was hit using `index.as_subindex(chunk)`,
3. and then get the index of the data we are reading/writing to/from the chunk using `chunk.as_subindex(chunk)`

As far as I can tell this PR does a few things:

1. It wraps those calls up into a single convenient function so we don't need to keep repeating ourselves
2. It saves time by not calling `as_subindex` for dimensions that the index doesn't hit
3. It saves time because calling `as_subchunks` is roughly equivalent to

```python
    def as_subchunks(self, idx, shape):
       for c in self.indices(shape):
           try:
               index = idx.as_subindex(c)
           except ValueError:
               # as_subindex raises ValueError in some cases when the
               # indices do not intersect (see the docstring of
               # as_subindex())
               continue

           if not index.isempty(self):
               # Yield those c for which idx.as_subindex(c) is nonempty
               yield c
```